docs/add_state_app.md

```diff
@@ -0,0 +1,252 @@
+# Add State to your app
+
+## What is State?
+
+We define access to a Streamlit app in a browser tab as a **session**. For each browser tab that connects to the Streamlit server, a new session is created. Streamlit reruns your script from top to bottom every time you interact with your app. Each reruns takes place in a blank slate: no variables are shared between runs.
+
+Session State is a way to share variables between reruns, for each user session. In addition to the ability to store and persist state, Streamlit also exposes the ability to manipulate state using Callbacks.
+
+In this guide, we will illustrate the usage of **Session State** and **Callbacks** as we build a stateful Counter app.
+
+For details on the Session State and Callbacks API, please refer to our [Session State API Reference Guide](session_state_api.md).
+
+## Build a Counter
+
+Let's call our script `counter.py`. It initializes a `count` variable and has a button to increment the value stored in the `count` variable:
+
+```python
+import streamlit as st
+
+st.title('Counter Example')
+count = 0
+
+increment = st.button('Increment')
+if increment:
+    count += 1
+
+st.write('Count = ', count)
+```
+
+No matter how many times we press the **_Increment_** button in the above app, the `count` remains at 1. Let's understand why:
+
+- Each time we press the **_Increment_** button, Streamlit reruns `counter.py` from top to bottom, and with every run, `count` gets initialized to `0` .
+- Pressing **_Increment_** subsequently adds 1 to 0, thus `count=1` no matter how many times we press **_Increment_**.
+
+As we'll see later, we can avoid this issue by storing `count` as a Session State variable. By doing so, we're indicating to Streamlit that it should maintain the value stored inside a Session State variable across app reruns.
+
+Let's learn more about the API to use Session State.
+
+### Initialization
+
+The Session State API follows a field-based API, which is very similar to Python dictionaries:
+
+```python
+# Check if 'foo' already exists in session_state
+# If not, then initialize it
+if 'foo' not in st.session_state:
```

Comment:
In general, I'm not a fan of `foo` and `bar`, as this jargon can be unfamiliar to new users. In my opinion, it would be better to do _anything_ else.

Reply:
Really good point. I will replace `foo` and `bar`